Add original item to restore plugin interface

mwieczorek opened the pull request:
Fixes #1092

ncdc commented on pkg/plugin/proto/RestoreItemAction.proto:
```diff
@@ -7,6 +7,7 @@ message RestoreExecuteRequest {
     string plugin = 1;
     bytes item = 2;
     bytes restore = 3;
+    bytes originalitem = 4;
```
Please capitalize the `I` in `originalItem`

ncdc replied:
@nrb @skriss in other bits of the restore code, we've called something like this `fromBackup` - how do you feel about that vs `originalItem`?

nrb replied:
I think `fromBackup` is a little clearer here

mwieczorek replied:
changed to `fromBackup`

ncdc commented on pkg/restore/restore.go:
```diff
@@ -627,6 +627,7 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
 	for _, file := range files {
 		fullPath := filepath.Join(resourcePath, file.Name())
 		obj, err := ctx.unmarshal(fullPath)
+		originalObj := obj.DeepCopy()
```
Please move this down below the error check

mwieczorek replied:
fixed

mwieczorek commented:
@ncdc Thanks for a quick response. 
I fixed the name of the parameter (`fromBackup`) and other small issues. Also, I added a file to the changelog folder.

BTW: I tested those changes with a simple plugin: https://github.com/mwieczorek/ark-plugins-test

ncdc commented on changelogs/unreleased/1123-mwieczorek:
```diff
@@ -0,0 +1 @@
+Add original resource (from backup) to restore plugin interface
```
```suggestion
Add original resource (from backup) to restore item action interface
```

mwieczorek replied:
fixed

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,7 @@ type ItemAction interface {
 	// the item from being restored) or error (which will be logged and will prevent the item
 	// from being restored) if applicable.
-	Execute(obj runtime.Unstructured, restore *api.Restore) (res runtime.Unstructured, warning error, err error)
+	Execute(obj runtime.Unstructured, fromBackup runtime.Unstructured, restore *api.Restore) (res runtime.Unstructured, warning error, err error)
```
This has 3 inputs and 3 outputs. My PR #1125 adds a 4th output. Maybe we need to consider switching to a single input struct and an output struct + error. WDYT? cc @carlisia

ncdc replied:
We should do this for backup item actions too, but we can do it in a separate PR.

The `warning` output variable makes things a bit messy here. We could do something like this:

```go
type RestoreItemActionExecuteOutput struct {
  UpdatedItem runtime.Unstructured
  Warning error
  Error error
}
```

but then the signature becomes `Execute(...) RestoreItemActionExecuteOutput`, with the `error` inside the struct instead of as the last output variable.

We could do this:

```go
type RestoreItemActionExecuteOutput struct {
  UpdatedItem runtime.Unstructured
  Warning error
}
```

along with `Execute(...) (RestoreItemActionExecuteOutput, error)`. Does that look weird or does it seem better?

mwieczorek replied:
Indeed the `warning` makes it a little bit messy, but I can live with that. So I prefer option 2. (it will also remove warning about two errors in golint). Also, maybe others from the team would like to suggest something?
Let me know so I can make changes (but after holidays, if you prefer to do it in your PR - go ahead and I'll rebase on it)

mwieczorek replied:
Added another commit with proposed changes. 
I had to tweak a little bit some tests (f.e. https://github.com/heptio/ark/pull/1123/files#diff-4cce89ebc50495fa33e9d2c373e87e6fR94) - and as I'm a golang newbie there may be a better way to do it - let me know

carlisia replied:
Sorry for the delay. Putting the multiple args in a struct is a great way to go. I too prefer the second option.

Unfortunately the code that this code calls has all been restructured (PR #1147), but we're going to have to go thru some merging pains, all good.

ncdc commented on pkg/restore/mocks/item_action.go:
```diff
@@ -1,19 +1,5 @@
-/*
```
Please don't remove the license header 😄

mwieczorek replied:
sorry, it wasn't me - it was mockery ;)

ncdc commented on pkg/restore/restore.go:
```diff
@@ -632,6 +632,8 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
 			continue
 		}
 
+		fromBackup := obj.DeepCopy()
```
Mind adding a comment above with a brief blurb about why we're doing the copy?

mwieczorek replied:
comment added

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+	Execute(input *RestoreItemActionExecuteInput) (*RestoreItemActionExecuteOutput, error)
+}
+
+// RestoreItemActionExecuteInput is a struct passed to Execute function
```
I would keep the type doc succinct and add per-field comments.
```go
// RestoreItemActionExecuteInput contains the input parameters for the ItemAction's Execute function.
```

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+// It contains item being processed by actions, unchanged item taken from backup
+// and restore spec
+type RestoreItemActionExecuteInput struct {
+	Item           runtime.Unstructured
```
```go
// Item is the item being restored. It is likely different from the pristine backed up version (metadata reset, changed by various restore item action plugins, etc.).
```

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+	Restore        *api.Restore
+}
+
+// RestoreItemActionExecuteOutput is a struct returned from Execute function
```
`RestoreItemActionExecuteOutput contains the output variables for the ItemAction's Execution function.`

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+// and restore spec
+type RestoreItemActionExecuteInput struct {
+	Item           runtime.Unstructured
+	ItemFromBackup runtime.Unstructured
```
Add field comment

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+type RestoreItemActionExecuteInput struct {
+	Item           runtime.Unstructured
+	ItemFromBackup runtime.Unstructured
+	Restore        *api.Restore
```
Add field comment

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+// RestoreItemActionExecuteOutput is a struct returned from Execute function
+// It contains updated item and possibly a warning message
+type RestoreItemActionExecuteOutput struct {
+	UpdatedItem runtime.Unstructured
```
Add field comment

ncdc commented on pkg/restore/item_action.go:
```diff
@@ -34,7 +34,23 @@ type ItemAction interface {
+// It contains updated item and possibly a warning message
+type RestoreItemActionExecuteOutput struct {
+	UpdatedItem runtime.Unstructured
+	Warning     error
```
Add field comment

ncdc commented on pkg/plugin/proto/RestoreItemAction.proto:
```diff
@@ -7,6 +7,7 @@ message RestoreExecuteRequest {
     string plugin = 1;
     bytes item = 2;
     bytes restore = 3;
+    bytes fromBackup = 4;
```
I think `itemFromBackup` is probably a bit clearer - wdyt?

ncdc commented on pkg/restore/job_action.go:
```diff
@@ -38,21 +36,21 @@ func (a *jobAction) AppliesTo() (ResourceSelector, error) {
 	}
 
-	return obj, nil, nil
+	return &RestoreItemActionExecuteOutput{input.Item, nil}, nil
```
WDYT about this (no warning)

```go
return NewRestoreItemActionExecuteOutput(input.Item), nil
```

and this (with warning)

```go
NewRestoreItemActionExecuteOutput(input.Item).WithWarning(someError)
```

ncdc commented on pkg/restore/job_action_test.go:
```diff
@@ -124,10 +124,10 @@ func TestJobActionExecute(t *testing.T) {
 			action := NewJobAction(arktest.NewLogger())
 
-			res, _, err := action.Execute(test.obj, nil)
+			res, err := action.Execute(&RestoreItemActionExecuteInput{test.obj, test.obj, nil})
```
Please use named fields

ncdc commented on pkg/restore/pod_action_test.go:
```diff
@@ -169,17 +169,18 @@ func TestPodActionExecute(t *testing.T) {
 			}
 
-			assert.Equal(t, test.expectedRes, res)
+			if res != nil {
```
I think it's stronger/safer to assert/require that res is non-nil when test.expectedErr is false?

ncdc commented on pkg/restore/restore.go:
```diff
@@ -632,6 +632,10 @@ func (ctx *context) restoreResource(resource, namespace, resourcePath string) (a
 		}
 
+		// make a copy of object retrieved from backup
+		// to make it available unchanged inside restore action
```
action -> actions

mwieczorek commented:
@ncdc I think I addressed all above comments. PTAL

mwieczorek commented:
@ncdc can you take a look at my fixes? and let me know if I can do anything else. thanks

ncdc commented:
@mwieczorek sorry, it's been a crazy past couple of weeks. I'll get back to this now.

ncdc commented:
I just opened #1175 to deal with the year issue in generated code. Once it's merged, @mwieczorek would you mind rerunning `make update` and pushing the changes?

ncdc commented on pkg/plugin/restartable_delegate_test.go:
```diff
@@ -88,6 +88,14 @@ func runRestartableDelegateTests(
 					}
 
+					// If function returns nil as struct return type, we cannot just
+					// compare the interface to nil as it's type will not be nil,
```
it's -> its

ncdc commented on pkg/restore/pod_action_test.go:
```diff
@@ -169,17 +169,23 @@ func TestPodActionExecute(t *testing.T) {
 
-			assert.Equal(t, test.expectedRes, res)
+			if !test.expectedErr {
+				assert.NotNil(t, res)
```
I'd to `require` instead of `assert` here

ncdc commented on pkg/restore/pod_action_test.go:
```diff
@@ -169,17 +169,23 @@ func TestPodActionExecute(t *testing.T) {
+			})
 
 			if test.expectedErr {
 				assert.NotNil(t, err, "expected an error")
```
We could add a `continue` here, and then we wouldn't need the `else` that follows (but we'd keep the assert-no-error check), and we also wouldn't need to check `if !test.expectedErr` again below.

mwieczorek replied:
adding `continue` tells me `you're not in a loop` ;)
I added `return`

mwieczorek commented:
rebased and pushed, PTAL and let me know if I should squash the commits etc

ncdc commented:
Thanks for keeping on top of this. We'll working on the renaming effort, but we should have spare cycles again next week. Thanks for being patient!

mwieczorek commented:
@ncdc Hi, just checking out with this. 
Should I rebase and resolve all conflicts or there's still some work around plugins and renaming which I should wait for?

ncdc commented:
@carlisia @skriss would you mind working with @mwieczorek on this?

carlisia commented:
I'm taking a look.

carlisia commented:
@mwieczorek the `plugins` branch has been updated with the renaming work we did. You may squash and rebase things. Thank you.

mwieczorek commented:
@carlisia Squashed and rebased.

carlisia commented:
Let's wait for one more 👍

mwieczorek commented:
@skriss - could you take a look?

skriss commented:
Yes, I'll look at this tomorrow.

skriss commented on pkg/restore/pod_action_test.go:
```diff
@@ -169,17 +170,23 @@ func TestPodActionExecute(t *testing.T) {
+			assert.Nil(t, err, "expected no error, got %v", err)
 
-			assert.Equal(t, test.expectedRes, res)
+			if !test.expectedErr {
```
this `if` can now be dropped since you're returning early if `test.expectedErr == true`

mwieczorek replied:
`if` removed

mwieczorek commented:
> Also - it doesn't look like we're actually using the `Warning` field returned from a restore item action anywhere - are we? If not, I'd motion to remove it as part of this branch.

Do you want to remove it completely from the interface?

skriss commented:
@mwieczorek we can do that in a separate PR -- I meant the shared `plugins` branch, not your branch for this PR.

skriss commented:
I added #1230 to capture that. This PR LGTM1!
